web/src/main/java/org/openmrs/web/filter/GZIPRequestStream.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.web.filter;
+
+import org.springframework.http.HttpEntity;
+
+import javax.servlet.ServletInputStream;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.http.HttpServletRequest;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.zip.GZIPInputStream;
+
+/**
+ * Wraps Response Stream for GZipFilter
+ * 
+ * @author Matt Raible
+ * @version $Revision: 1.3 $ $Date: 2004/05/16 02:17:00 $
+ */
+public class GZIPRequestStream extends ServletInputStream {
+	
+	// abstraction of the output stream used for compression
+	protected OutputStream bufferedOutput = null;
+	
+	// state keeping variable for if close() has been called
+	protected boolean closed = false;
+	
+	// reference to original response
+	protected HttpServletRequest request = null;
+	
+	// reference to the output stream to the client's browser
+	protected ServletOutputStream output = null;
+	
+	protected GZIPInputStream zipInput = null;
+	
+	// reference to the input stream
+	protected ServletInputStream input = null;
+	
+	// default size of the in-memory buffer
+	private int bufferSize = 50000;
+	
+	//    /**
+	//     * Gets the input stream from a response entity. If the entity is gzipped then this will get a stream over the uncompressed data.
+	//     * @param entity the entity whose content should be read
+	//     * @return the input stream to read from
+	//     * @throws IOException
+	//     */
+	//    public static ServletInputStream getUngzippedContent(ServletInputStream responseStream) throws IOException {
+	//        if (responseStream == null) {
+	//            return responseStream;
+	//        }
+	//        if(contentEncoding.contains("gzip"))   responseStream=new GZIPInputStream(responseStream);
+	//        return responseStream;
+	//    }
+	
+	public GZIPRequestStream(HttpServletRequest request) throws IOException {
+		super();
+		closed = false;
+		this.request = request;
+		this.input = request.getInputStream();
+		this.zipInput = new GZIPInputStream(input);
+		
+		bufferedOutput = new ByteArrayOutputStream();
+	}
+	
+	public int read() throws IOException {
+		return input.read();
```

Comment:
java.io.EOFException: Unexpected end of ZLIB input stream
	at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:240)
	at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158)
	at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:117)
	at org.openmrs.web.filter.GZIPRequestStream.read(GZIPRequestStream.java:77)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.readLine(BufferedReader.java:324)
	at java.io.BufferedReader.readLine(BufferedReader.java:389)
	at org.openmrs.web.filter.update.GZIPFilterTest$1.matches(GZIPFilterTest.java:78)
	at org.mockito.internal.matchers.LocalizedMatcher.matches(LocalizedMatcher.java:30)
	at org.mockito.internal.invocation.ArgumentsComparator.argumentsMatch(ArgumentsComparator.java:26)
	at org.mockito.internal.invocation.ArgumentsComparator.argumentsMatch(ArgumentsComparator.java:18)
	at org.mockito.internal.invocation.InvocationMatcher.matches(InvocationMatcher.java:58)
	at org.mockito.internal.invocation.InvocationsFinder$RemoveNotMatching.isOut(InvocationsFinder.java:145)
	at org.mockito.internal.invocation.InvocationsFinder$RemoveNotMatching.isOut(InvocationsFinder.java:137)
	at org.mockito.internal.util.collections.ListUtil.filter(ListUtil.java:16)
	at org.mockito.internal.invocation.InvocationsFinder.findInvocations(InvocationsFinder.java:20)
	at org.mockito.internal.verification.checkers.MissingInvocationChecker.check(MissingInvocationChecker.java:32)
	at org.mockito.internal.verification.Times.verify(Times.java:36)
	at org.mockito.internal.verification.MockAwareVerificationMode.verify(MockAwareVerificationMode.java:21)
	at org.mockito.internal.handler.MockHandlerImpl.handle(MockHandlerImpl.java:72)
	at org.mockito.internal.handler.NullResultGuardian.handle(NullResultGuardian.java:29)
	at org.mockito.internal.handler.InvocationNotifierHandler.handle(InvocationNotifierHandler.java:38)
	at org.mockito.internal.creation.MethodInterceptorFilter.intercept(MethodInterceptorFilter.java:51)
	at $javax.servlet.FilterChain$$EnhancerByMockitoWithCGLIB$$5de2954d.doFilter(<generated>)
	at org.openmrs.web.filter.update.GZIPFilterTest.zipRequestTest(GZIPFilterTest.java:69)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:74)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:82)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:72)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:240)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:180)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:74)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:211)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:67)